### PR TITLE
Orch api factoring

### LINKED ISSUE #9207: stub out the Orchestrator API
## What is the Problem Being Solved?

https://github.com/agoric-labs/orchestration-api-spec/ describes a spec for an Orchestrator helper object. We should validate it by implementing it in agoric-sdk. Not all the pieces are available yet but we can fill out much of it to ensure it fits.

## Description of the Design

Make the Exo class with tests. Whatever doesn't yet work, assert that they throw.

Include JSDoc and Markdown documentation so partners can see how they would use it, so that they can give feedback.

### PR DESCRIPTION
**Pull request by turadg**
closes: #9207

## Description

Refactoring the API types and perhaps implementation to satsisfy:
- a types module specifying the public chain-agnostic interface for Orchestration
  - it must not expose Cosmos SDK types
  - it should not change
- a types module for the IBC subclass of the public interface

### CODE REVIEW

File: packages/orchestration/src/orchestration-api.d.ts
Comment (dtribble) [inline packages/orchestration/src/orchestration-api.d.ts]: Why is this not just "query"?  Putting ICQ in the name is very specific and narrow (and doesn't account for e.g., v1 vs v2).

File: packages/orchestration/src/orchestration-api.d.ts
Comment (dtribble) [inline packages/orchestration/src/orchestration-api.d.ts]: If it's not a known chain, do we have a generic "KnownChain" for "unknown"?

File: packages/orchestration/src/orchestration-api.d.ts
Comment (dtribble) [inline packages/orchestration/src/orchestration-api.d.ts]: I was expecting to see OrchestrationChain and perhaps local chain in here as well. Where are those?

File: packages/orchestration/src/orchestration-api.d.ts
Comment (turadg) [reply packages/orchestration/src/orchestration-api.d.ts]: It snuck in at some point. I suppose because https://github.com/agoric-labs/orchestration-api-spec/blob/main/src/orchestration.d.ts didn't have a query method.

If we do give it `query()`, what are the types it takes and returns? Is it conditioned on the particular chain? Do we create a query abstraction over all chains?

File: packages/orchestration/src/orchestration-api.d.ts
Comment (turadg) [reply packages/orchestration/src/orchestration-api.d.ts]: That hasn't moved (this is draft) because it wasn't entirely clean of Cosmos types. E.g. its `makeAccount` turns an `OrchestrationAccount` which has delegation methods that use Cosmos types. Maybe they just need to move out of the generic interface and down into a Cosmos delegation one

File: packages/cosmic-proto/package.json
Comment (0xpatrickdev) [inline packages/cosmic-proto/package.json]: Out of scope for this PR? I suspect the removal of these might break some tests

**Review by 0xpatrickdev — APPROVED** [review]
LGTM. Considering we paired on this, I suspect you may also want a ✅ from @dtribble before merging

File: packages/cosmic-proto/package.json
Comment (turadg) [reply packages/cosmic-proto/package.json]: these are all redundant with the subpatterns above them

Comment (turadg) [reply packages/cosmic-proto/package.json]: hmm, you're right tho. [breaking](https://github.com/Agoric/agoric-sdk/actions/runs/9038366546/job/24839392140?pr=9356). Oh, I forgot our bundler doesn't support wildcards yet!

I'll repush without this change

**Comment by turadg** [comment]
Taking out of draft. There's one lint failure but it's on `no-use-before-define` and I'm making a separate PR to disable that rule,
- https://github.com/Agoric/agoric-sdk/pull/9361

File: packages/orchestration/src/types.ts
Comment (dckc) [inline packages/orchestration/src/types.ts:8]: this looks circular. I can imaging that's not a ts error, but it seems unhelpful for thinking about things.

File: packages/orchestration/src/orchestration-api.ts
Comment (dckc) [inline packages/orchestration/src/orchestration-api.ts]: well, it depends on `./types.js`, which depends on Cosmos types such as `MsgBeginRedelegateResponse` by way of `cosmos-api`.

File: packages/orchestration/src/types.ts
Comment (dckc) [inline packages/orchestration/src/types.ts]: This code ends up in vats, right? Let's not use typescript in vat code until we have a build process that does nothing more than removing types.

_Closest issue I can find is #5760._

**Review by dckc — DISMISSED** [review]
Let's talk more before going to .ts for vat code.

File: packages/orchestration/src/types.ts
Comment (turadg) [reply packages/orchestration/src/types.ts]: This module has no runtime. If it did, the bundle would fail because it sources from SCM and not the package dist

File: packages/orchestration/src/orchestration-api.ts
Comment (turadg) [reply packages/orchestration/src/orchestration-api.ts]: Are you finding that the types exported by this module depend on Cosmos types? I didn't yet attempt to make the entire modules layered, just the APIs they describe.

File: packages/orchestration/src/types.ts
Comment (turadg) [reply packages/orchestration/src/types.ts:8]: Is this a blocking request for the PR? I can untangle the circularity but I wouldn't want to do that until the API surface is agreed upon

File: packages/orchestration/src/cosmos-api.ts
Comment (dtribble) [inline packages/orchestration/src/cosmos-api.ts]: This we would like to NOT be cosmos-specific.

Comment (dtribble) [reply packages/orchestration/src/cosmos-api.ts]: I have heard that Gravity has an encoding of eth etc. denoms.

File: packages/orchestration/src/cosmos-api.ts
Comment (dtribble) [inline packages/orchestration/src/cosmos-api.ts]: Also should be non-cosmos

File: packages/orchestration/src/cosmos-api.ts
Comment (dtribble) [reply packages/orchestration/src/cosmos-api.ts]: https://github.com/PeggyJV/gravity-bridge/blob/main/module/x/gravity/types/ethereum.go

I think they use `gravity/ethcontractadress` but that doesn't include the Chain ID.  So we would want something more like `eth/<chainId>/<contractAddress>`. The key is that it fits into the Denom formats, so we can just use Denom everywhere.

File: packages/orchestration/src/cosmos-api.ts
Comment (dtribble) [inline packages/orchestration/src/cosmos-api.ts]: Am I correct that this is unchanged because you are just moving things rather than changing them in this PR

File: packages/orchestration/src/cosmos-api.ts
Comment (dtribble) [inline packages/orchestration/src/cosmos-api.ts]: Registry?  Is htis a new thing or a renaming?

File: packages/orchestration/src/cosmos-api.ts
Comment (dtribble) [inline packages/orchestration/src/cosmos-api.ts]: Should be generic.

File: packages/orchestration/src/cosmos-api.ts
Comment (dtribble) [reply packages/orchestration/src/cosmos-api.ts]: Also, "Coin" may imply too much specifics, which is why we used "ChainAmount".  Is this being renamed in this PR

File: packages/orchestration/src/examples/unbondExample.contract.js
Comment (dtribble) [inline packages/orchestration/src/examples/unbondExample.contract.js]: I think we want the staking operations on the account object.  It's certainly as legit as `liquidStake` as a custom operation on the stride account

File: packages/orchestration/src/orchestration-api.ts
Comment (dtribble) [inline packages/orchestration/src/orchestration-api.ts:57]: do we use "ethereum" or just "eth"?   I'm thinking "eth" is actually better because it's a format used on lots of chains.  Alternatively, "evm", but that woudl surprise most people.

File: packages/orchestration/src/orchestration-api.ts
Comment (dtribble) [inline packages/orchestration/src/orchestration-api.ts]: I think that some things beginning with Chain and some with Coin would be odd.

File: packages/orchestration/src/types.ts
Comment (dtribble) [inline packages/orchestration/src/types.ts]: why a type.ts file rather than type.d.ts?

File: packages/orchestration/src/types.ts
Comment (dckc) [reply packages/orchestration/src/types.ts]: oops. I misread. I missed the `type` in `export type ...`.

File: packages/orchestration/src/types.ts
Comment (dtribble) [inline packages/orchestration/src/types.ts]: Or
```
...ICAOperations,
...cosmosOperations
```
which would include ICA and staking and such.

Comment (dtribble) [reply packages/orchestration/src/types.ts]: Balance checks and the like seem like they should be in a shared set of operations.

File: packages/orchestration/src/types.ts
Comment (dckc) [reply packages/orchestration/src/types.ts:8]: > Is this a blocking request for the PR?

no.

I'm trying to understand the proposal.

File: packages/orchestration/src/types.ts
Comment (dtribble) [inline packages/orchestration/src/types.ts]: shoudl have most of the methods of celestia?

File: packages/orchestration/src/types.ts
Comment (dtribble) [inline packages/orchestration/src/types.ts]: Omit is magic. What's this mean?

File: packages/orchestration/src/orchestration-api.ts
Comment (dckc) [reply packages/orchestration/src/orchestration-api.ts]: a TODO would make it more clear.

It's a `@file` comment, and it says "**must** not depend on Cosmos type **in anyway**".

File: packages/orchestration/src/types.ts
Comment (dtribble) [inline packages/orchestration/src/types.ts]: this should have all the ICA and staking operations?

File: packages/orchestration/src/types.ts
Comment (dtribble) [inline packages/orchestration/src/types.ts]: Nit: Order of "agoric type | low level type" should match between DenomArg and AmountArg.

Comment (dtribble) [reply packages/orchestration/src/types.ts]: For usability, I expect the low-level type should be first.

File: packages/orchestration/src/types.ts
Comment (dckc) [reply packages/orchestration/src/types.ts]: TIL: `.d.ts` are taken as axioimatically correct by `tsc`.

@turadg explained that this is written in `.ts` so it gets checked.

_I wonder if that's worth a comment for other readers. hm._

File: packages/orchestration/src/types.ts
Comment (turadg) [reply packages/orchestration/src/types.ts]: How is it magic?

https://www.typescriptlang.org/docs/handbook/utility-types.html#omittype-keys

File: packages/orchestration/src/orchestration-api.ts
Comment (0xpatrickdev) [reply packages/orchestration/src/orchestration-api.ts:57]: `hex` or `hexadecimal` might be a good option. It seems like three possible values here could be `bech32`, `base58`, and `hex` - I spiked a little on this in [this gist](https://gist.github.com/0xpatrickdev/8fdd882a232cbae3d4d65e1eb4403a81).

File: packages/orchestration/src/types.ts
Comment (turadg) [reply packages/orchestration/src/types.ts]: because a `.d.ts` file has two downsides,
1. requires an empty `.js` file next to it
2. prevents type checking of the file contents (because tsc trusts `.d.ts` as its own output)

I'll try to document this justification

Comment (turadg) [reply packages/orchestration/src/types.ts]: https://github.com/Agoric/agoric-sdk/pull/9364/

File: packages/orchestration/src/cosmos-api.ts
Comment (turadg) [reply packages/orchestration/src/cosmos-api.ts]: new thing, a place to get by info by Denom, which was taken to be Cosmos only. No longer needed with Denom being a universal namespace

File: packages/orchestration/src/cosmos-api.ts
Comment (turadg) [reply packages/orchestration/src/cosmos-api.ts]: yes, by the logic that it was Cosmos-only, and that's what Cosmos types call it.

I'm putting it back to `ChainAmount`, but I'll note the mismatch with Agoric's "amount" that aren't all fungible.

it also seems to conflate amount naming with chain naming, whereas on Agoric there are many amounts for the one chain.

WDYT of `DenomAmount` or `NativeAmount`?  Or perhaps to reduce confusion with the `Amount` type, `DenomValue` or `DenomCount`.

Comment (turadg) [reply packages/orchestration/src/cosmos-api.ts]: actually, `NativeAmount` would be easily confused with _Agoric-native_ (ERTP) Amount.
```
/** Amounts can be provided as pure data using denoms or as native Amounts */
export type AmountArg = ChainAmount | Amount;
```
maybe `RemoteAmount` to convey the difference. Though that still sounds like a _kind_ of (ERTP) `Amount`.

I see the goal of naming these as like ERTP Amounts, but I think having "Amount" in the name of this not ERTP Amount is more confusing than helpful because:
- it doesn't work in AmountMath
- it can't be used in Amount parameters

File: packages/orchestration/src/cosmos-api.ts
Comment (turadg) [reply packages/orchestration/src/cosmos-api.ts]: Generic on what? `Amount` has a generic slot for Kind but there's nothing generic about that on ChainAmount; it's always 'nat'. I'll add that to the return type.

Are you looking for a mapping of the argument's "denom" to the return value's "brand"? That would require a generic slot on Amount, which I've toyed with but is out of scope of this PR. Is that something you want queued up?

File: packages/orchestration/src/cosmos-api.ts
Comment (turadg) [reply packages/orchestration/src/cosmos-api.ts]: yes

File: packages/orchestration/src/cosmos-api.ts
Comment (turadg) [reply packages/orchestration/src/cosmos-api.ts]: Also having it be an Amount made `asAmount` confusing because it's ostensibly already an amount. That led me to omit Amount from the name. I'm pushing with `TokenCount` which conveys both that it's fungible and not an ERTP Amount.

```js
asAmount: (amount: TokenCount) => NatAmount;
```

File: packages/orchestration/src/cosmos-api.ts
Comment (0xpatrickdev) [reply packages/orchestration/src/cosmos-api.ts]: I think we might have better luck with an object here. You are right to call out that for ethereum (and solana) the important identifiers are `contractAddress` and `chainId` - `denom`s do not come into play until someone decides to make a representation on an IBC chain.

Aside - other evm-interop chains do not follow the same `baseDenom` pattern as gravity . On axelar they look something like: `dai-wei`, `dot-planck`, `uusdc`, `weth-wei`. And on kava: `erc20/tether/usdt`. I think we'll want to preserve these values in the `denom` name space, and use additional keys (`contractAddress`, `chainId`) to facilitate building messages for non-IBC chains.

Comment (dtribble) [reply packages/orchestration/src/cosmos-api.ts]: The general that we don't need to commit a particular, I don't think I would want an object for all of these, since we get a lot of them back that are well known in context. All those fit just fine in the model of string encoding. these are just arms that someone on their chain can decide to use. Everyone else uses indirect references to them.

File: packages/orchestration/src/cosmos-api.ts
Comment (dtribble) [reply packages/orchestration/src/cosmos-api.ts]: Oops. I meant "generic" as in applying to all chains and not just cosmos.

File: packages/orchestration/src/orchestration-api.ts
Comment (dtribble) [reply packages/orchestration/src/orchestration-api.ts:57]: I like "hex" except that it's hex for a public key in the specific algorithm, etc. But if there's a representation name for eth keys, that would be better than both.

File: packages/orchestration/src/types.ts
Comment (dtribble) [reply packages/orchestration/src/types.ts]: Seems more clear to construct types additively rather than with subtraction. So what is the name of the type that doesn't include those operations?

File: packages/orchestration/src/types.ts
Comment (turadg) [reply packages/orchestration/src/types.ts]: I wrote a TODO instead of tackling this now

File: packages/orchestration/src/types.ts
Comment (turadg) [reply packages/orchestration/src/types.ts]: I wrote an XXX instead of tackling this now. I expect these will be autogenerated

File: packages/orchestration/src/types.ts
Comment (turadg) [reply packages/orchestration/src/types.ts]: I read this as a the request:

>  "agoric type | low level type"

but you want it the other way? I agree it's better to have low level type first, but I want to confirm your request before I make another edit

File: packages/orchestration/src/types.ts
Comment (turadg) [reply packages/orchestration/src/types.ts]: fair. Agoric is one special case among Cosmos chains so I don't think it needs its own type alias, but I'll change it to Pick instead of Omit

File: packages/orchestration/src/orchestration-api.ts
Comment (0xpatrickdev) [reply packages/orchestration/src/orchestration-api.ts:57]: Yes `hex` doesn't fully capture it. It seems to be the last 20 bytes of the Keccak-256 hash of the pubkey. Maybe it should be `keccak256`?

Regarding public keys, we could include [pubKeyType](https://gist.github.com/0xpatrickdev/8fdd882a232cbae3d4d65e1eb4403a81#file-address-d-ts-L28-L32): `secp256k1 ` | `ed25519` | `sr25519` and `pubKey`.  However, for ICAs (a type of `ModuleAccount`) and contract-controlled accounts on evm chains, there doesn't seem to be a concept of pubkey.  Including these might also indicate we are leveraging cryptographic keys for authn versus the application layer.

_OTOH, I can foresee some scenarios where we might need to go between bech32<>hex for some cosmos<>ethereum interop. If talking to the evm side of chains like dydx, injective, and evmos, we'll likely need to use hex encoding. Currently unclear to me if pubkeys are required to go from bech32->hex, or if a bech32Prefix is sufficent._

File: packages/orchestration/src/cosmos-api.ts
Comment (dtribble) [reply packages/orchestration/src/cosmos-api.ts]: We agreed on DenomAmount

File: packages/orchestration/src/types.ts
Comment (dtribble) [reply packages/orchestration/src/types.ts]: use info so it looks like a localhost connection. Eventually it may be replaced with that implementation.

**Review by dtribble — APPROVED** [review]
Turadg and I walked through comments and agreed on the results. I consider it auto-approved after he makes those changes.

**Pull request merged**
